src/node_messaging.cc

```diff
@@ -589,12 +590,19 @@ void MessagePort::OnMessage() {
       // Call the JS .onmessage() callback.
       HandleScope handle_scope(env()->isolate());
       Context::Scope context_scope(context);
-      Local<Value> args[] = {
-        received.Deserialize(env(), context).FromMaybe(Local<Value>())
-      };
 
-      if (args[0].IsEmpty() ||
-          MakeCallback(env()->onmessage_string(), 1, args).IsEmpty()) {
+      Local<Object> event;
+      Local<Value> payload;
+      Local<Value> cb_args[1];
+      if (!received.Deserialize(env(), context).ToLocal(&payload) ||
+          !env()->message_event_object_template()->NewInstance(context)
+              .ToLocal(&event) ||
+          event->Set(context, env()->data_string(), payload).IsNothing() ||
+          event->Set(context, env()->target_string(), object()).IsNothing() ||
+          (cb_args[0] = event, false) ||
```

Comment:
Can this be assigned right away at the declaration of `cb_args`?

`Local<Value> cb_args[1] = { event };`

Reply:
@lundibundi At that point `event` is still uninitialized…

I could try to refactor the code so that we don’t have this chained-if here. The downside would be that making sure the error path is executed gets a bit more complex, I think.

Reply:
Oh, sorry, forgot about copy constructor call here.

This 'if' is indeed complicated but I don't see a good way of making it better.